sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
+
+            /* Check if there are broadcasts to send and transmit immediately */
+            if(packet_queue_length(&(lwmac.tx.neighbours[0].queue)) > 0) {
+                lwmac.tx.current_neighbour = &(lwmac.tx.neighbours[0]);
+                lwmac_set_state(TRANSMITTING);
+                break;
+            }
+
+            lwmac_tx_neighbour_t* neighbour = _next_tx_neighbour(&lwmac);
+
+            if(neighbour != NULL) {
+                /* Offset in microseconds when the earliest (phase) destination
+                 * node wakes up that we have packets for. */
+                int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+                /* If there's not enough time to prepare a WR to catch the phase
+                 * postpone to next interval */
+                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+                }
+
+                time_until_tx -= LWMAC_WR_PREPARATION_US;
+                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+                /* Register neighbour to be the next */
+                lwmac.tx.current_neighbour = neighbour;
+
+                /* Stop dutycycling, we're preparing to send. This prevents the
+                 * timeout arriving late, so that the destination phase would
+                 * be missed. */
```

Comment:
> Procedure-B

That's gonna be it. At the end of the rtt interrupt,
"sched_context_switch_request" is set because of the sent message,
causing an immediate context switch to the LwMAC thread.

On 06/22/2016 03:59 PM, zhuoshuguo wrote:
> In sys/net/gnrc/link_layer/lwmac/lwmac.c
> <https://github.com/RIOT-OS/RIOT/pull/3730#discussion_r68057392>:
> 
>> +
>> +                /* If there's not enough time to prepare a WR to catch the phase
>> +                 * postpone to next interval */
>> +                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
>> +                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
>> +                }
>> +
>> +                time_until_tx -= LWMAC_WR_PREPARATION_US;
>> +                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
>> +
>> +                /* Register neighbour to be the next */
>> +                lwmac.tx.current_neighbour = neighbour;
>> +
>> +                /* Stop dutycycling, we're preparing to send. This prevents the
>> +                 * timeout arriving late, so that the destination phase would
>> +                 * be missed. */
> 
> Thanks! I understand that "The message sending is neccessary", what I
> wonder about is that:
> 
>     "After rtt_cb() has finished, execution will continue in the idle
>     thread."
> 
> for me, the tricky thing is that if the message hasn't reached the LwMAC
> thread in time (thus to trigger |lpm_prevent_sleep |= LWMAC_LPM_MASK|)
> before the idle thread execution, then, since now idle thread is
> running, then, by judging the current |lpm_prevent_sleep|, why not does
> the idle thread stop the CPU again? (before the execution of
> |lpm_prevent_sleep |= LWMAC_LPM_MASK| in LwMAC thread.)
> 
> To make me clear: which is the process procedure in RIOT:
> procedure-A
> 1) RTT alarm arrives, |rtt_cb()| be called and CPU is waked up (all
> threads wake up), |rtt_cb()| sends message to LwMAC thread;
> 2) (before the message reach LwMAC thread, namely *message delivering
> execution*) since LwMAC has nothing to do, the execution of CPU goes to
> the *idle thread*;
> 3) by judging the current |lpm_prevent_sleep|, *idle thread* put the CPU
> into sleep mode. which is bad.
> 
> Or,
> Procedure-B
> 1) RTT alarm arrives, |rtt_cb()| be called and CPU is waked up (all
> threads wake up), |rtt_cb()| sends message to LwMAC thread;
> 2) the CPU *immediately execute the action of sending the message to
> LwMAC thread*,
> 3) then, right *after step-2)* (the message has arrived the message
> queue of LwMAC thread), the CPU executes the thread of LwMAC (due to
> higher priority over idle thread);
> 4) LwMAC executes |lpm_prevent_sleep |= LWMAC_LPM_MASK|, and then this
> totally wake up the CPU (I mean prevent the idle thread to set lmp mode).
> 
> You can see my question is about the execution timing and sequence. :-)
> Thanks!
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3730/files/8691130abd2b0dccfb0660c1776e4e74e6e9571f#r68057392>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEdn2JZTM3wHuZxa4qw8BcmNp4qf1lAYks5qOT-pgaJpZM4FzcvM>.
>